// ng-chartist/esm2020/lib/chartist.module.mjs
import { NgModule } from "@angular/core";
import { ChartistComponent } from "./chartist.component";
import * as i0 from "@angular/core";
export class ChartistModule {
}
ChartistModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ChartistModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ChartistModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ChartistModule, declarations: [ChartistComponent], exports: [ChartistComponent] });
ChartistModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ChartistModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ChartistModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ChartistComponent],
                    imports: [],
                    exports: [ChartistComponent],
                }]
        }] });
